src/Build/Evaluation/LazyItemEvaluator.UpdateOperation.cs

```diff
@@ -43,23 +46,69 @@ protected override void ApplyImpl(ImmutableList<ItemData>.Builder listBuilder, I
                     return;
                 }
 
-                ItemSpecMatchesItem matchItemspec;
-                bool? needToExpandMetadataForEachItem = null;
+                SetMatchItemSpec();
+                _itemsToUpdate ??= ImmutableList.CreateBuilder<ItemBatchingContext>();
+                _itemsToUpdate.Clear();
 
+                for (int i = 0; i < listBuilder.Count; i++)
+                {
+                    var itemData = listBuilder[i];
+
+                    var matchResult = _matchItemSpec(_itemSpec, itemData.Item);
+
+                    if (matchResult.IsMatch)
+                    {
+                        listBuilder[i] = UpdateItem(listBuilder[i], matchResult.CapturedItemsFromReferencedItemTypes);
+                    }
+                }
+
+                DecorateItemsWithMetadata(_itemsToUpdate.ToImmutableList(), _metadata, _needToExpandMetadataForEachItem);
+            }
+
+            internal ItemData UpdateItem(ItemData item)
```

Comment:
Definitely summarize this method since it's internal. 
A very rough idea:
"Set the item spec, do X with _itemsToUpdate, and return the cloned matching item"